query_optimizer/logical/Logical.hpp

```diff
@@ -64,12 +70,62 @@ class Logical : public OptimizerTree<Logical> {
    */
   virtual std::vector<expressions::AttributeReferencePtr> getReferencedAttributes() const = 0;
 
+  /**
+   * @brief Make a copy of this logical node with input expressions replaced by
+   *        input_expression.
+   *
+   * @return A copy of this logical node with the new input expressions.
+   */
+  virtual LogicalPtr copyWithNewInputExpressions(
+      const std::vector<expressions::ExpressionPtr> &input_expressions) const {
+    // Default implementation: return a shared reference to "this" when there is
+    // no change.
+    DCHECK(input_expressions.empty());
+    return shared_from_this();
+  }
+
+  /**
+   * @return The input expressions in this Logical.
+   */
+  const std::vector<expressions::ExpressionPtr>& input_expressions() const {
+    return input_expressions_;
+  }
+
+  /**
+   * @return The input attributes, which are the union set of the output attributes of
+   *         all children.
+   */
+  std::vector<expressions::AttributeReferencePtr> getInputAttributes() const {
+    std::vector<expressions::AttributeReferencePtr> input_attributes;
+    for (const LogicalPtr &child : children()) {
+      const std::vector<expressions::AttributeReferencePtr> child_output_attributes =
```

Comment:
I think we need to have an agreement on the semantics of `A a = b'. In other words, in your idea what is the output for the following C++ program?

```
#include <vector>
#include <iostream>
using namespace std;

class A {
 public:
  A() {
  }

  A(const A& a) {
    std::cerr << "Copy constructor\n";
  }

  A(A&& a) {
    std::cerr << "Move constructor\n";
  }

  A& operator= (const A& a) {
    std::cerr << "Copy assignment\n";
    return *this;
  }

  A& operator= (A&& a) {
    std::cerr << "Move assignment\n";
    return *this;
  }
};

A f() {
  A a;
  return a;
}

int main(int argc, char *argv[]) {
  A a;

  // 1
  A b = a;

  // 2
  A c = f();

  // 3
  A d = std::move(f());

  return 0;
}
```